Fixes: one-to-one hash wallet

jzaki opened the pull request:
## What is this PR doing?
Ensuring a wallet only has one bls key registered for it.

## How can these changes be manually tested?
@kautukkundan can you write a test for this?

## Does this PR resolve or contribute to any issues?
Fixes #104 

## Checklist
- [x] I have manually tested these changes
- [x] Post a link to the PR in the group chat

## Guidelines
- If your PR is not ready, mark it as a draft
- The `resolve conversation` button is for reviewers, not authors
  - (But add a 'done' comment or similar)

jacque006 commented on contracts/contracts/VerificationGateway.sol:
```diff
@@ -89,30 +89,31 @@ contract VerificationGateway
-                )
-            ))
-        )))));
-        if (!hasCode(walletAddress)) {
-            walletAddress = address(0);
-        }
-        return IWallet(payable(walletAddress));
-    }
+        // if (externalWalletsFromHash[hash] != IWallet(address(0))) {
+        //     return externalWalletsFromHash[hash];
+        // }
+
+        // address walletAddress = address(uint160(uint(keccak256(abi.encodePacked(
+        //     bytes1(0xff),
+        //     address(this),
+        //     hash,
+        //     keccak256(abi.encodePacked(
+        //         type(TransparentUpgradeableProxy).creationCode,
+        //         abi.encode(
+        //             address(blsWalletLogic),
+        //             address(walletProxyAdmin),
+        //             getInitializeData()
+        //         )
+        //     ))
+        // )))));
+    //     IWallet walletAddress = registeredWalletFromHash[hash];
+    //     if (!hasCode(address(walletAddress))) {
+    //         walletAddress = IWallet(address(0));
+    //     }
+    //     return walletAddress;
+    // }
```
1. Is there a reason to keep this code around? If so, might be better to move it into a docstring, other documentation, or a Solidity lib.
2. This will break [BlsWalletWrapper.Nonce](https://github.com/web3well/bls-wallet/blob/main/contracts/clients/src/BlsWalletWrapper.ts#L144), so we will need to update the client lib, publish a new version, and update consumers.

jzaki replied:
1. No, this is a preliminary commit
2. I don't think this should break anything since the deleted function interface `walletFromHash` is the same is the hash/key mapping, now of the same name. All smart contract tests pass unchanged

jzaki replied:
@kautukkundan as well as the specific tests for the fix, feel free to remove this too.

jacque006 reviewed (commented):
Overall core changes look good :+1:

kautukkundan commented on contracts/package.json:
```diff
@@ -17,6 +17,7 @@
     "@nomiclabs/hardhat-ethers": "^2.0.0",
     "@nomiclabs/hardhat-etherscan": "^2.1.3",
     "@nomiclabs/hardhat-waffle": "^2.0.0",
+    "@thehubbleproject/bls": "^0.5.1",
```
removed in 140f5e7, did not require any other change. the contract tests pass without this additional dependency.

voltrevo commented:
@jzaki @jacque006 I think we should generally be merging things as soon as we approve them rather than giving it back to the author to decide. Or maybe we need to talk about this policy?

(Is there a reason why this one wasn't merged last week?)

jzaki commented:
There are some things that have surfaced in my review that I'm in the process of addressing. Can share details once complete.

jzaki commented:
- Made room in the VerificationGateway by making proxy admin deployment external (requires setting address in constructor as per bls lib param)
- added a delay between wallet's setting of a new blskey hash (using one-to-one reverse mapping)
- removed bls key from wallet
- updated tests
Closes #104 #155
